README.md
# Catclass
An image classifier employing a deep neural network to identify pictures of
cats.

## Installation
[Clone the repo][orig] to your local machine. You will also need a copy of the
[Deepen][deep] library, which is not yet released to PyPI.
```(bash)
~/../catclass$ pip install ~/path/to/deepen
~/../catclass$ pip install -r requirements.txt
```

## Usage
```(bash)
~/../catclass$ python -m catclass
```
On my laptop (Intel i7 CPU @ 2.20GHz × 8; 16GB RAM), it takes about 35s to train
(209 examples) and test (50 examples) the model through the sample data set, with
training accuracy 98.6% and test accuracy 80.0% using the default parameters.

## Contributing
Catclass provides a safe, welcoming space for collaboration. Everyone
contributing to this project—including the codebase, issue trackers, chat, email,
social media and the like—is expected to uphold our [Code of Conduct][coc].

Bug reports and pull requests are welcome on [GitHub][orig].

To contribute code, first [fork the project][fork] on GitHub and make a local
clone. Create a topic branch, make and commit your changes, and push this
branch back to GitHub. [Submit a pull request][pull] and ask to have your work
reviewed for inclusion in the main repository.

## License
This project is available as open source under the terms of the [MIT License][mit].

---
_This file is composed with [GitHub Flavored Markdown][gfm]._

[coc]:  https://github.com/petejh/catclass/blob/master/CODE_OF_CONDUCT.md
[deep]: https://github.com/petejh/deepen
[fork]: https://help.github.co://help.github.com/en/github/getting-started-with-github/fork-a-repo
[gfm]:  https://github.github.com/gfm/
[orig]: https://github.com/petejh/catclass
[mit]:  https://github.com/petejh/catclass/blob/master/LICENSE.txt
[pull]: https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/creating-a-pull-request-from-a-fork/


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

This project uses [Semantic Versioning][sv].

## [Unreleased][new]

## [0.1.0][0.1.0] — 2020-04-24

### Added
- Add routines to load, normalize, and partition data from an H5 file into
  training and test sets.
- Parameterize specs for the data file.
- Add `train()` for training the CatClassifier.
- Add `test()` for testing and making predictions with the trained model.
- Add the CLI module for running the classifier from the command line.

### Changed
- Move CatClassifier.run() to CLI.run(). Running the package will train the
  model, displaying incremental cost and final accuracy for the default data set.

## [0.0.0][0.0.0] — 2020-04-05

### Added
- Create the project. An image classifier employing a deep neural network to
  identify pictures of cats.

---
_This file is composed with [GitHub Flavored Markdown][gfm]._

[gfm]: https://github.github.com/gfm/
[sv]: https://semver.org

[new]: https://github.com/petejh/catclass/compare/HEAD..v0.1.0
[0.1.0]: https://github.com/petejh/catclass/releases/tag/v0.1.0
[0.0.0]: https://github.com/petejh/catclass/releases/tag/v0.0.0


test/test_catclass.py
import unittest

import catclass

class TruthTest(unittest.TestCase):
    def test_it_knows_the_truth_when_it_sees_it(self):
        self.assertTrue(True)

    def test_it_knows_the_truth_when_it_does_not_see_it(self):
        self.assertTrue(False)


catclass/catclass.py
import os

import h5py as h5
import numpy as np

from deepen.model import Model

class CatClassifier:

    def __init__(self):
        self.model = Model()

    @property
    def layer_dims(self):
        return self.model.layer_dims

    @layer_dims.setter
    def layer_dims(self, layer_dims):
        self.model.layer_dims = layer_dims

    def train(self, train_x, train_y, iterations):
        params, costs = self.model.learn(train_x, train_y, iterations=iterations)

        return (params, costs)

    def test(self, test_x):
        test_pred = self.model.predict(test_x)

        return test_pred

    def load_data(self, data_spec):
        """Load datasets from a specified H5-format file.

        Parameters
        ----------
        data_spec: dict of {str: str}
            Specifications for the data file.

            datafile: Path to the data file relative to this script.
            x_name: Name of the input dataset.
            y_name: Name of the output dataset.
            class_name: Name of the classifications dataset.

        Returns
        -------
        tuple of numpy array

            data_x: The input dataset.
            data_y: The output dataset.
            data_class: The classifications dataset.
        """

        path = os.path.dirname(os.path.abspath((__file__)))
        datapath = os.path.join(path, data_spec['datafile'])

        with h5.File(datapath, 'r') as datafile:
            data_x = datafile[data_spec['x_name']][:]
            data_y = datafile[data_spec['y_name']][:]
            data_class = datafile[data_spec['class_name']][:]

        return (data_x, data_y, data_class)

    def normalize_data(self, data_x, data_y):
        """Reshape and normalize the given datasets.

        Parameters
        ----------
        data_x: numpy array of int in [0..255]
            The input dataset, having shape (num_examples, x, y, 3).
        data_y: numpy array
            The output dataset, having shape (num_examples, 1).

        Returns
        -------
        tuple of numpy array

            norm_x: numpy array of float in [0..1]
                The input dataset, having shape(x * y * 3, num_examples).
            norm_y: numpy array
                The output dataset, having shape (1, num_examples).
        """

        norm_x = data_x.reshape(data_x.shape[0], -1).T
        norm_x = norm_x / 255

        norm_y = data_y.reshape(1, data_y.shape[0])

        return (norm_x, norm_y)

    def partition_data(self, data_x, data_y, test_frac=0.193):
        """Partition the data into training and test sets.

        Parameters
        ----------
        data_x: numpy array
            The input dataset, having shape (:, num_examples).
        data_y: numpy array
            The output dataset, having shape (1, num_examples).
        test_frac: float in [0..1]
            The fraction of the total number of examples in the dataset that
            should be reserved for the test set.

        Returns
        -------
        tuple of numpy array

            train_x: Training set input data, having shape (:, num_examples - test_examples).
            train_y: Training set output data, having shape (1, num_examples - test_examples).
            test_x: Test set input data, having shape (:, test_examples).
            test_y: Test set output data, having shape (1, test_examples).
        """

        test_cases = int(data_x.shape[1] * test_frac + 0.5)
        test_index = data_x.shape[1] - test_cases

        train_x = data_x[:, 0:test_index]
        train_y = data_y[:, 0:test_index]
        test_x = data_x[:, test_index:]
        test_y = data_y[:, test_index:]

        return (train_x, train_y, test_x, test_y)


catclass/cli.py
from time import time
from datetime import timedelta

import numpy as np

import catclass

class CLI:

    @staticmethod
    def run():
        classifier = catclass.CatClassifier()

        data_spec = {
            'datafile': 'data/data.h5',
            'x_name': 'data_x',
            'y_name': 'data_y',
            'class_name': 'data_class'
        }
        data_x, data_y, data_class = classifier.load_data(data_spec)

        norm_x, norm_y = classifier.normalize_data(data_x, data_y)
        train_x, train_y, test_x, test_y = classifier.partition_data(norm_x, norm_y)

        assert(train_x.shape == (12288, 209))
        assert(train_y.shape == (1, 209))
        assert(test_x.shape == (12288, 50))
        assert(test_y.shape == (1, 50))

        np.random.seed(1)
        classifier.layer_dims = [norm_x.shape[0], 20, 7, 5, 1]
        iterations = 2500

        print("Training model...", end='')
        start = time()
        params, costs = classifier.train(train_x, train_y, iterations=iterations)
        end = time()
        print("done.")
        print("Time to train: %s" %  str(timedelta(seconds=(end - start))))

        for i in range(iterations):
            if i % 100 == 0:
                print("Cost at iteration %i: %f" % (i, costs[i]))

        train_pred = classifier.test(train_x)
        train_accuracy = np.sum(train_pred == train_y) / train_y.shape[1]
        print("Training accuracy: %3.3f" % train_accuracy)

        test_pred = classifier.test(test_x)
        test_accuracy = np.sum(test_pred == test_y) / test_y.shape[1]
        print("Test accuracy: %3.3f" % test_accuracy)
